src/deploy/functions/params.ts

```diff
@@ -267,83 +396,271 @@ export async function resolveParams(
-          return res;
+      return promptText<number>(prompt, param.input, resolvedDefault, (res: string) => {
+        if (isNaN(+res)) {
+          return { message: `"${res}" could not be converted to a number.` };
+        }
+        if (res.includes(".")) {
+          return { message: `${res} is not an integer value.` };
         }
-        logger.error(`${param.label || param.name} must be an integer; retrying...`);
+        return +res;
+      });
+  }
+}
+
+async function promptResourceString(
+  prompt: string,
+  input: ResourceInput,
+  projectId: string,
+  resolvedDefault?: string
+): Promise<string> {
+  const notFound = new FirebaseError(`No instances of ${input.resource.type} found.`);
+  switch (input.resource.type) {
+    case "storage.googleapis.com/Bucket":
+      const buckets = await listBuckets(projectId);
+      if (buckets.length === 0) {
+        throw notFound;
       }
+      const forgedInput: SelectInput<string> = {
+        type: "select",
+        select: buckets.map((bucketName: string): SelectOptions<string> => {
+          return { label: bucketName, value: bucketName };
+        }),
+      };
+      return promptSelect<string>(prompt, forgedInput, resolvedDefault, (res: string) => res);
+    default:
+      logger.warn(
+        `Warning: unknown resource type ${input.resource.type}; defaulting to raw text input...`
+      );
+      return promptText<string>(prompt, { type: "text" }, resolvedDefault, (res: string) => res);
+  }
+}
+
+type retryInput = { message: string };
+async function promptText<T extends RawParamValue>(
+  prompt: string,
+  input: TextInput<T>,
+  resolvedDefault: T | undefined,
+  converter: (res: string) => T | retryInput
+): Promise<T> {
+  const res = await promptOnce({
+    type: "input",
+    default: resolvedDefault,
+    message: prompt,
+  });
+  if (input.validationRegex) {
+    const userRe = new RegExp(input.validationRegex);
+    if (!userRe.test(res)) {
+      logger.error(
+        input.validationErrorMessage ||
+          `Input did not match provided validator ${userRe.toString()}, retrying...`
+      );
+      return promptText<T>(prompt, input, resolvedDefault, converter);
+    }
   }
+  const converted = converter(res);
+  if (typeof converted === "object") {
+    logger.error(converted.message);
+    return promptText<T>(prompt, input, resolvedDefault, converter);
+  }
+  return converted;
+}
+
+async function promptSelect<T extends RawParamValue>(
+  prompt: string,
+  input: SelectInput<T>,
+  resolvedDefault: T | undefined,
+  converter: (res: string) => T
+): Promise<T> {
+  const response = await promptOnce({
+    name: "input",
+    type: "list",
+    default: () => {
+      resolvedDefault;
+    },
+    message: prompt,
+    choices: input.select.map((option: SelectOptions<T>): ListItem => {
+      return {
+        checked: false,
+        name: option.label,
+        value: option.value.toString(),
+      };
+    }),
+  });
+  return converter(response);
```

Comment:
q: why is it safe to not worry about "conversion gone wrong" here?

Reply:
Ostensibly the types were correct when the developer defined the select input, but Daniel is right. Some jerk could try an int param with a select dialog that return "foo". We might as well have an error message if the select option's value can't be coerced into the param's type. Parsing is the worst.